server/sonar-server/src/main/java/org/sonar/server/settings/ws/SetAction.java

```diff
@@ -126,6 +146,64 @@ public void handle(Request request, Response response) throws Exception {
     response.noContent();
   }
 
+  private void doHandle(DbSession dbSession, SetRequest request) {
+    Optional<ComponentDto> component = searchComponent(dbSession, request);
+    checkPermissions(component);
+
+    PropertyDefinition definition = propertyDefinitions.get(request.getKey());
+
+    if (!request.getFieldValues().isEmpty()) {
+      doHandlePropertySet(dbSession, request, definition, component);
+    } else {
+      validate(request, component);
+      dbClient.propertiesDao().insertProperty(dbSession, toProperty(request, component));
+      if (!component.isPresent()) {
+        settingsChangeNotifier.onGlobalPropertyChange(request.getKey(), persistedValue(request));
+      }
+    }
+  }
+
+  private void doHandlePropertySet(DbSession dbSession, SetRequest request, @Nullable PropertyDefinition definition, Optional<ComponentDto> component) {
+    validatePropertySet(request, definition, component);
+
+    int[] fieldIds = IntStream.rangeClosed(1, request.getFieldValues().size()).toArray();
+    String inlinedFieldKeys = IntStream.of(fieldIds).mapToObj(String::valueOf).collect(COMMA_JOINER);
+    String key = persistedKey(request);
+    Long componentId = component.isPresent() ? component.get().getId() : null;
+
+    deleteSettings(dbSession, component, key);
+    dbClient.propertiesDao().insertProperty(dbSession, new PropertyDto().setKey(key).setValue(inlinedFieldKeys).setResourceId(componentId));
+
+    List<Map<String, String>> fieldValues = request.getFieldValues();
+    IntStream.of(fieldIds).boxed()
+      .flatMap(i -> fieldValues.get(i - 1).entrySet().stream().map(entry -> new KeyValue(key + "." + i + "." + entry.getKey(), entry.getValue())))
+      .forEach(keyValue -> dbClient.propertiesDao().insertProperty(dbSession, toFieldProperty(keyValue, componentId)));
+  }
+
+  private void deleteSettings(DbSession dbSession, Optional<ComponentDto> component, String key) {
+    if (component.isPresent()) {
+      settingsUpdater.deleteComponentSetting(dbSession, key, component.get());
+    } else {
+      settingsUpdater.deleteGlobalSetting(dbSession, key);
+    }
+  }
+
+  private void validatePropertySet(SetRequest request, @Nullable PropertyDefinition definition, Optional<ComponentDto> component) {
+    checkRequest(definition != null, "Setting '%s' is undefined", request.getKey());
+    checkRequest(PropertyType.PROPERTY_SET.equals(definition.type()), "Parameter '%s' is used for setting of property set type only", PARAM_FIELD_VALUES);
+    checkComponentQualifier(request, definition, component);
```

Comment:
Anyway, UT needs to be adding to cover the usage of property set settings only available on global | on some component qualifiers.